src/Kestrel.Core/KestrelServerOptions.cs

```diff
@@ -55,6 +57,31 @@ public class KestrelServerOptions
         /// </summary>
         public KestrelServerLimits Limits { get; } = new KestrelServerLimits();
 
+        /// <summary>
+        /// Provides a configuration source where endpoints will be loaded from on server start.
+        /// The default is null.
+        /// </summary>
+        public IKestrelConfigurationBuilder ConfigurationBuilder { get; set; }
+
+        /// <summary>
+        /// A default configuration action for all endpoints. Use for Listen, configuration, the default url, and URLs.
+        /// </summary>
+        internal Action<ListenOptions> EndpointDefaults { get; set; } = _ => { };
+
+        /// <summary>
+        /// Used to flow settings for connection adapters and other extensions.
+        /// </summary>
+        public IDictionary<string, object> AdapterData { get; } = new Dictionary<string, object>(0);
+
+        /// <summary>
+        /// Specifies a configuration Action to run for each newly created endpoint.
+        /// </summary>
+        /// <param name="configureOptions"></param>
+        public void ConfigureEndpointDefaults(Action<ListenOptions> configureOptions)
+        {
+            EndpointDefaults = configureOptions ?? throw new ArgumentNullException(nameof(configureOptions));
```

Comment:
Should we allow this to stack? Either way the doc comments should mention whether this overrides previous calls to ConfigureEndpointDefaults.

Reply:
I don't think more that one default action is required. Comments updated.